### PR TITLE
Onboard users to App Hosting using Developer Connect

### PR DESCRIPTION
**Pull request by mathu97**
### Description

Enables app hosting onboarding using developer connect. This path is gated by a flag `--with-dev-connect` or `-w`. The changes in this PR should not affect or change the current flow of onboarding using Cloud Build repositories.

### Sample Commands

FIREBASE_DEVELOPERCONNECT_P4SA_URL=<p4sa_url> CLOUD_SECRET_MANAGER_URL=<secret_manager_url> FIREBASE_DEVELOPERCONNECT_URL=<dev_connect_url> firebase apphosting:backends:create --project=<firebase_project> --with-dev-connect

**Note**: This flow is currently only being tested with Developer Connect autopush. Staging and Prod envs are not ready yet.

### CODE REVIEW

File: src/commands/apphosting-backends-create.ts [hunk 1]
```diff
@@ -13,11 +13,27 @@ export const command = new Command("apphosting:backends:create")
     "specify the service account used to run the server",
     "",
   )
+  .option(
```
Comment (mathu97) [inline src/commands/apphosting-backends-create.ts:16]: add test

File: src/commands/apphosting-connections-list.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { Command } from "../command";
```
Comment (mathu97) [inline src/commands/apphosting-connections-list.ts:1]: move to a separate PR (this is only helpful for testing)

File: src/commands/apphosting-connections-reset.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { Command } from "../command";
```
Comment (mathu97) [inline src/commands/apphosting-connections-reset.ts:1]: move to a separate PR (this is only helpful for testing)

File: src/commands/index.ts [hunk 4]
```diff
@@ -165,6 +165,9 @@ export function load(client: any): any {
     client.apphosting.backends = {};
     client.apphosting.backends.list = loadCommand("apphosting-backends-list");
     client.apphosting.backends.create = loadCommand("apphosting-backends-create");
+    client.apphosting.connections = {};
```
Comment (mathu97) [inline src/commands/index.ts:168]: move to a separate PR (this is only helpful for testing)

File: src/init/features/apphosting/index.ts [hunk 5]
```diff
@@ -36,6 +39,105 @@ const apphostingPollerOptions: Omit<poller.OperationPollerOptions, "operationRes
...
+
+  await Promise.all([
+    ensure(projectId, developerConnectOrigin(), "apphosting", true),
+    ensure(projectId, cloudbuildOrigin(), "apphosting", true),
+    ensure(projectId, secretManagerOrigin(), "apphosting", true),
+    ensure(projectId, cloudRunApiOrigin(), "apphosting", true),
+    ensure(projectId, artifactRegistryDomain(), "apphosting", true),
+  ]);
+
+  logBullet("All required APIs are enabled");
```
Comment (mathu97) [inline src/init/features/apphosting/index.ts:60]: remove

File: src/gcp/devConnect.ts [hunk 6]
```diff
@@ -147,6 +170,7 @@ export async function getConnection(
   location: string,
   connectionId: string,
 ): Promise<Connection> {
+  location = LOCATION_OVERRIDE;
```
Comment (mathu97) [inline src/gcp/devConnect.ts:173]:
```suggestion
  location = LOCATION_OVERRIDE ?? location
```
Move LOCATION_OVERRIDE to env variable in `api.ts`. The `LOCATION_OVERRIDE` can be removed later, but it is  helpful for debugging dev connect in different environments.

File: src/init/features/apphosting/githubConnections.ts [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+  }
+  const { projectId, location, id } = match.groups as unknown as ConnectionNameParts;
+  return {
+    projectId,
+    location,
+    id,
+  };
+}
+
+export function isApphostingConnection(name: string): boolean {
```
Comment (mathu97) [inline src/init/features/apphosting/githubConnections.ts:50]: move to separate PR

File: src/init/features/apphosting/index.ts [hunk 8]
```diff
@@ -351,3 +453,39 @@ export async function orchestrateRollout(
   }
   return { rollout, build };
 }
+
+/**
+ * lists dev connect connections
+ */
+export async function listDeveloperConnectAppHostingConnections(
```
Comment (mathu97) [inline src/init/features/apphosting/index.ts:460]: move to separate Pr

File: src/init/features/apphosting/index.ts [hunk 9]
```diff
@@ -351,3 +453,39 @@ export async function orchestrateRollout(
...
+  for (let i = 0; i < appHostingConnections.length; i++) {
+    const connection = appHostingConnections[i];
+    logBullet(connection.name);
+  }
+}
+
+/**
+ * deletes all dev connect connections
+ */
+export async function resetDeveloperConnectAppHostingConnections(
```
Comment (mathu97) [inline src/init/features/apphosting/index.ts:477]: move to separate PR

File: src/init/features/apphosting/index.ts [hunk 10]
```diff
@@ -36,6 +39,105 @@ const apphostingPollerOptions: Omit<poller.OperationPollerOptions, "operationRes
   masterTimeout: 25 * 60 * 1_000,
   maxBackoff: 10_000,
 };
+/**
+ * Set up a new App Hosting backend.
+ */
+export async function doSetupWithDeveloperConnect(
```
Comment (mathu97) [inline src/init/features/apphosting/index.ts:45]: might be better to use the original `doSetup` function with a flag to use dev connect instead of cloud build

File: src/init/features/apphosting/githubConnections.ts [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+        `\t  --role="roles/secretmanager.admin\n`,
+    );
+    throw new FirebaseError("Insufficient IAM permissions to create a new connection to GitHub");
+  }
+
+  try {
+    await rm.addServiceAccountToRoles(projectId, dcsaEmail, ["roles/secretmanager.admin"], true);
+  } catch (e: any) {
+    // if the dev connect P4SA doesn't exist in the project, generate one
+    if (e?.code === 400) {
```
Comment (mathu97) [inline src/init/features/apphosting/githubConnections.ts:300]: Add test coverage for this

**Review by tonyjhuang — CHANGES_REQUESTED** [review]

File: src/gcp/devConnect.ts [hunk 12]
```diff
@@ -3,7 +3,14 @@ import { developerConnectOrigin, developerConnectP4SAOrigin } from "../api";
 import { generateServiceIdentity } from "./serviceusage";
 
 const PAGE_SIZE_MAX = 1000;
-const LOCATION_OVERRIDE = "us-central1"; // TODO(mathusan): for debugging only
+
+/**
+ * LOCATION_OVERRIDE is used to test Developer Connect in different environments.
+ *
+ * Autopush -> 'us-central1'
+ * Staging -> 'us-west1'
```
Comment (tonyjhuang) [inline src/gcp/devConnect.ts:11]: I wonder if we should include this information, I don't see any other references to autopush/staging envs in this repo

Comment (tonyjhuang) [reply src/gcp/devConnect.ts:11]: Also are we planning on removing this after we launch? How is it different from calling `backends:create --location`?

**Review by tonyjhuang — CHANGES_REQUESTED** [review]
Almost there :)

File: src/init/features/apphosting/githubConnections.ts [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+      projectId,
+      dcsaEmail,
+      ["roles/secretmanager.admin"],
+      /* skipAccountLookup= */ true,
+    );
+  } catch (e: any) {
+    // if the dev connect P4SA doesn't exist in the project, generate one
+    if (e?.code === 400 || e?.status === 400) {
+      await devConnect.generateP4SA(projectNumber);
+      await rm.addServiceAccountToRoles(
```
[the remark this answers is not in the document]
Comment (mathu97) [reply src/init/features/apphosting/githubConnections.ts:300]: > You're right in that calling GenerateServiceIdentity multiple times after the P4SA already exists will always return a done operation (eg from your testing), but this may not be the case for a new onboarding project that doesn't already have a p4sa.

No I meant on the first call. I've tested it a couple times, creating a new project from scratch. On first call to GenerateServiceIdentity we get an operation with the name `finished.DONE_OPERATION ` with `done = true`. And polling with that operation name results in a 400 error. I can add the polling logic and skip polling if the returned operation is already done.

File: src/gcp/devConnect.ts (hunk 12 above)
Comment (mathu97) [reply src/gcp/devConnect.ts:11]: > I wonder if we should include this information, I don't see any other references to autopush/staging envs in this repo
ya I think it might cause some confusion. I'll remove. 

> Also are we planning on removing this after we launch? How is it different from calling backends:create --location?

Yes this will be removed as soon as developer connect production is good to go. The reason for this is because we can only create backends in `us-central1` at the moment, and if we were to test in dev connect in staging we need to use `us-west1` because dev connect staging is only spun up in that location.

**Review by tonyjhuang — APPROVED** [review]
